Implements the navigator page api

chunhtai opened the pull request:
## Description

This is part 2 of https://github.com/flutter/flutter/issues/45938
Design doc: https://flutter.dev/go/navigator-with-router

This pr adds the page api, transition delegate and modifies route entry lifecycle


## Related Issues

https://github.com/flutter/flutter/issues/45938

## Tests

I added the following tests:

A lot tests is navigator_test.dart

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I signed the [CLA].
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] All existing and new tests are passing.
- [ ] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [ ] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [ ] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: *Replace with a link to your migration guide*

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

chunhtai committed:
Implements the navigator page api

chunhtai committed:
more good stuffs

chunhtai commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -888,14 +1266,69 @@ class Navigator extends StatefulWidget {
+  /// contain the [Page] for the given [Route]. The next time the [pages] list
+  /// is updated, if the [Page] corresponding to this [Route] is still present,
+  /// it will be interpreted as a new route to display.
+  final OnPopPageCallback onPopPage;
```
Do you think we should have some kind of notifier for remove and complete as well?

chunhtai replied:
I see, so the philosophy is to discourage mix usage of page api and imperative api unless it is a pageless route or legendary android back bottom

chunhtai commented:
bump for review

goderbauer commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -1991,8 +2475,290 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
+    int oldEntriesTopToScan = oldEntriesTop;
+    final Map<LocalKey, _RouteEntry> pageKeyToOldRoute = <LocalKey, _RouteEntry>{};
+    final Map<Route<dynamic>, List<Route<dynamic>>> pageRouteToPagelessRoutes =
+      <Route<dynamic>, List<Route<dynamic>>>{};
```
nit: indent by 4

goderbauer replied:
Also: probably best to just keep this on one line.

goderbauer commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -1991,8 +2475,290 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
+        pagelessRoutes.add(oldEntry.route);
+        continue;
+      }
+      // The oldEntry should be a page route at this point.
```
express this as an assert?

goderbauer commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -1991,8 +2475,290 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
+      // The oldEntry should be a page route at this point.
+      previousPageRoute = oldEntry.route;
+      final Page<dynamic> page = previousPageRoute.settings as Page<dynamic>;
+      // Page with key will be removed regardless.
```
What do you mean by this comment?

goderbauer commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -1991,8 +2475,290 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
+    }
+    // Updates the middle of the list.
+    //
+    // The idea is to find two pages in the new page list that has corresponding
```
.. that have ...

goderbauer commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -1991,8 +2475,290 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
+        // it into the history.
+        final _RouteEntry newEntry = _RouteEntry(
+          nextPage.createRoute(context),
+          initialState: _RouteLifecycle.idle,
```
idle doesn't seem like the right state here.

goderbauer commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -1991,8 +2475,290 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
+            continue;
+
+          final Page<dynamic> potentialPageToRemove =
+          potentialEntryToRemove.route.settings as Page<dynamic>;
```
put this on the previous line or indent by 4

goderbauer commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -1991,8 +2475,290 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
+            if (
+              !pageKeyToNewPage.containsKey(potentialPageToRemove.key) ||
+              !potentialEntryToRemove.canUpdateFrom(pageKeyToNewPage[potentialPageToRemove.key])
+            )
```
put {} around the if branch to make it easier to read.

goderbauer commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -1991,8 +2475,290 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
+      }
+      if (enteringPageRoutes.isNotEmpty || exitingPageRoutes.isNotEmpty) {
+        final Iterable<Route<dynamic>> results = widget.transitionDelegate
+          ._transition(
```
leave `._transition` on the previous line.

goderbauer commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -1991,8 +2475,290 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
+            enteringPageRoutes: enteringPageRoutes,
+            exitingPageRoutes: exitingPageRoutes,
+            precedingRoutes: newHistory.map((_RouteEntry entry) => entry.route)
+              .toList(),
```
leave this on the previous line

goderbauer commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -1991,8 +2475,290 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
+            exitingPageRoutes: exitingPageRoutes,
+            precedingRoutes: newHistory.map((_RouteEntry entry) => entry.route)
+              .toList(),
+            succeedingRoutes: _history.sublist(oldEntriesTop).map((
```
put the last ( on the next line.

goderbauer commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -1991,8 +2475,290 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
+      addPageRoute(matchingEntry.route);
+    }
+
+    // Any remaining old routes that does not have a match will need to be
```
... that do not have ...

goderbauer commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -1991,8 +2475,290 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
+
+    if (enteringPageRoutes.isNotEmpty || exitingPageRoutes.isNotEmpty) {
+      final Iterable<Route<dynamic>> results = widget.transitionDelegate
+        ._transition(
```
formatting here just like above.

goderbauer commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -1991,8 +2475,290 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
+    // Verifies we either reach the bottom or the oldEntriesTop must be updatable
+    // by newPagesTop.
+    assert(() {
+      if (oldEntriesTop <= oldEntriesBottom)
```
here and elsewhere: please use {} around if and else branches.

goderbauer commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -2001,7 +2767,7 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
     _RouteEntry entry = _history[index];
     _RouteEntry previous = index > 0 ? _history[index - 1] : null;
-    bool canRemove = false;
+    bool canRemoveOrAdd = false; // Whether there is a fully opaque route on top to silent remove or add route underneath.
```
... to silently ...

goderbauer commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -2011,12 +2777,22 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
+              navigator: this,
+              previous: previous?.route,
+              previousPresent: _getRouteBefore(
+                index - 1, _RouteEntry.isPresentPredicate)?.route,
```
Just leave this on one line.

goderbauer commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -2630,7 +3411,13 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
+      if (widget.onPopPage(entry.route, result))
+        entry.currentState = _RouteLifecycle.pop;
+    } else {
+      // A pageless route will need to be popped manually.
```
Maybe remove this comment? Seems confusing.

goderbauer reviewed (commented):
Haven't fully reviewed the pageUpdate algorithm and the tests yet.

chunhtai committed:
addressing part of comments

chunhtai commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -876,14 +1254,69 @@ class Navigator extends StatefulWidget {
+  /// route and is tied to the [Route] that _does_ correspond to a [Page] object
+  /// that is below it in the history.
+  ///
+  /// Pages that are removed (and any routes that were pushed over those pages
```
This paragraph is removing a page based route and its corresponding pageless route can animate based on tranistiondelegate. How would add factor into this paragraph?

chunhtai commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -1728,19 +2161,24 @@ class Navigator extends StatefulWidget {
 // # These states await futures or other events, then transition automatically.
 enum _RouteLifecycle {
-  // routes that are and will be present:
-  add, // we'll want to run install, didAdd, etc; a route created by onGenerateInitialRoutes or by the initial widget.pages
```
yes, however add/adding is not considered present after page api. In Navigator we need to determine:
1, which route is considered present to user. This is willBePresent. This is used when annoucing previous route after multiple pop or decide if a route can receive user gesture.
2, which route is considered present to navigator. This is isPresent. The Navigator use this decide which route still worth notifying navigation change
3, which route is still animatable, whether or not navigator cares about them. This is canTransit. These routes still need to be chained up with didchangeprev and didchangenext so the secondary animations are still wired it up.

When a page is add/adding, it is waiting for the top most route push to finish. In this case, this page is consider not exist in the app yet and not suitable for transition as well. Thus it is rated below removing

chunhtai replied:
`add` will go to `adding` synchronically. If there is no top pushing route, it will go to `idle` synchronically. If a route is `add` or `adding` it means it is waiting for top pushing route and should not be displayed and thus non transition-able. 

it will actually create some issue if we consider it transition-able

Consider this case, we have routes [A, B, C]. If we put D, E, F on top, it become [A, B, C ,D(add), E(add), F(push)].
Navigator will try to announce F.didChangePrev(E)  and  E.didChangeNext(F) to hook up secondary animation because it think E is transition-able. This will result C's secondary animation never hooked up with F and will not transition with F

chunhtai replied:
talked offline, this will also affect some imperative api to determine whether a route can be popped. Need to refactor it

chunhtai commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -1728,19 +2161,24 @@ class Navigator extends StatefulWidget {
   replace, // we'll want to run install, didReplace, etc; a route added via replace() and friends
   idle, // route is being harmless
-  // routes that are but will not present:
+  // routes that are not present to the users, but the navigator state still consider they are present.
```
will update the document

chunhtai commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -1929,20 +2386,34 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
   @override
   void initState() {
     super.initState();
+    assert(
```
Asserting a property of a list property will make it a non const constructor. Am I missing something?

chunhtai replied:
That sounds fine, but i would still have to assert if the list empty in initstate and didUpdateWidget

chunhtai replied:
actually that will be the same thing as what i am doing now, which is the following except i didn't check onPop.
```
class Foo {
  const Foo(this.list, Function onPop) 
    : assert(list == null || onPop != null), 
      _list = list ?? const [];
  
  final list;
}
```

goderbauer commented:
Cirrus seems pretty unhappy.

chunhtai committed:
fix cirrus

goderbauer commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -1991,8 +2598,241 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
+      pageKeyToOldEntry[page.key] = oldEntry;
+    }
+
+    // Scans middle of the new pages and records page key to new page map.
```
What's this for?

goderbauer commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -1991,8 +2598,241 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
+    while (newPagesBottomToScan <= newPagesTop) {
+      final Page<dynamic> page = widget.pages[newPagesBottomToScan];
+      if (page.key != null) {
+        assert(!pageKeyToNewPage.containsKey(page.key), 'Duplicated page key.');
```
Do we need to check for duplicated keys also in the beginning and end of the list that we matched above?

goderbauer commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -1991,8 +2598,241 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
+        final _RouteEntry matchingEntry = pageKeyToOldEntry.remove(nextPage.key);
+        assert(matchingEntry.canUpdateFrom(nextPage));
+        matchingEntry.route._updateSettings(nextPage);
+        newHistory.add(matchingEntry);
```
What about the pageless routes above matchingEntry?

chunhtai replied:
the new history only contains page-based route all the pageless route will be inserted back after transition delegate resolves them

goderbauer commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -1991,8 +2598,241 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
+      }
+    }
+
+    // Any remaining old routes that do not have a match will need to be removed.
```
Why does this require another loop? Aren't those just the ones from pageKeyToOldEntry?

goderbauer replied:
pageRouteToPagelessRoutes could be updated in one of the previous loops through the middle, no?

goderbauer replied:
Similarly, the locationToExitingPageRoute could just be build while we can the middle and when we reuse a route we remove it from there.

chunhtai replied:
How do we efficiently remove the route been reused from locationToExitingPageRoute? we would either need to do a linear search or have another map to reverse lookup the location, or record the map in reversed way and re-reverse the map at the end. Space O(2n) X Time O(n)   or Space(n) X TimeO(2n)

goderbauer commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -1991,8 +2598,241 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
+    }());
+
+    // Updates the top of the list.
+    while ((oldEntriesBottom <= oldEntriesTop) && (newPagesBottom <= newPagesTop)) {
```
Why can we not do this directly in the second while loop of this function?

chunhtai replied:
if we do that we will have to store it into a list in reverse order and them combine at the end. It need to iterate through them twice anyway. This method will save more space, and this is actually following the same algorithm as updateChildren

goderbauer commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -1991,8 +2598,241 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
+      oldEntriesBottom += 1;
+      newPagesBottom += 1;
+    }
+    final Iterable<_RouteEntry> results = widget.transitionDelegate._transition(
```
nit: add a blank line above this one since it doesn't belong to the previous block anymore.

chunhtai commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -505,6 +620,273 @@ class NavigatorObserver {
 
+/// An [Route] wrapper interface that can be staged for [TransitionDelegate] to
+/// decide how its underlying [Route] should transit on or off screen.
+abstract class StageableRoute{
```
This will create some confusing. _RouteEntry extends this class. If the class StagedRoute, does that imply _RouteEntry is always Staged?

Hixie replied:
This isn't a Route, so its name shouldn't end in Route.

Maybe RouteEntry? Or RouteTransitionRecord?

chunhtai commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -505,6 +620,273 @@ class NavigatorObserver {
+        assert(!exitingPageRoute._debugIsExiting);
+        if (pageRouteToPagelessRoutes.containsKey(exitingPageRoute)) {
+          for (final StageableRoute pagelessRoute in pageRouteToPagelessRoutes[exitingPageRoute]) {
+            assert(!pagelessRoute._debugIsExiting);
```
if it is a decision has been made, this field will no longer be true, I should update the name...

chunhtai commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -1728,61 +2229,78 @@ class Navigator extends StatefulWidget {
+  bool get hasPage => route.settings is Page;
+
+  bool canUpdateFrom(Page<dynamic> page) {
+    if (currentState.index > _RouteLifecycle.idle.index)
```
nice catch, it should be able to..

chunhtai committed:
addressing review comments

chunhtai commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -461,6 +500,82 @@ class RouteSettings {
+///
+/// The type argument `T` is the corresponding [Route]'s return type, as
+/// used by [Route.currentResult], [Route.popped], and [Route.didPop].
+class PageBuilder<T> extends Page<T> {
```
I am thinking how useful is this builder. We are not surfacing page api, and it is likely the Router will take care of changing route name string to configured navigator.

chunhtai committed:
wording

chunhtai committed:
fix test

chunhtai committed:
fix route announcement

chunhtai commented:
ready for another review @goderbauer

goderbauer commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -1896,6 +2468,14 @@ class _RouteEntry {
+    return currentState.index <= _RouteLifecycle.removing.index &&
+      currentState.index >= _RouteLifecycle.push.index;
+  }
+  bool get canTransitition {
```
What does "canTransition" mean? Why is this limited to lifecycles between remove and push?

chunhtai replied:
I will update the name, this field should return true for route that are suitable for transition animation. I limited the lifecycle between remove and push because add/adding is not suitable for transition animation because it has to wait for top most route to finish pushing animation

goderbauer commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -1991,8 +2653,276 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
+    }());
+
+    // This attempts to diff the new pages list (widget.pages) with
+    // the old _RouteEntry[s] list (_history), and produce a new list of
```
produces

goderbauer commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -1991,8 +2653,276 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
+    // Page-based routes without keys might be synced but there is no guarantee.
+
+    // The general approach is to sync the entire new list backwards, as follows:
+    // 1. Walk the lists from the top, syncing nodes, and records pageless routes,
```
record

goderbauer commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -1991,8 +2653,276 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
+    //    until you no longer have matching nodes.
+    // 2. Walk the lists from the bottom, without syncing nodes, until you no
+    //    longer have matching nodes. We'll sync these nodes at the end. We
+    //    don't sync them now because we want to sync all the nodes in order
```
Why do we need to sync them in order?

chunhtai replied:
I wrote this document other way around... top should be bottom and bottom should be top.

the second step walk the lists from the top in reverse order. If we sync them now, it will be stored in reverse order as well. we have to reverse them again to merge the results back to new history. It will be O(2n) anyway. however, if we sync them now, we will have to spend another O(n) space to store the reversed list and it does not worth.

goderbauer commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -1991,8 +2653,276 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
+    // 3. Walk the narrowed part of the old list to get the list of
+    //    keys.
+    // 4. Walk the narrowed part of the new list forwards:
+    //     * Creates a new _RouteEntry for non-keyed items and records them for
```
Create

goderbauer replied:
record

goderbauer commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -1991,8 +2653,276 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
+    //     * Creates a new _RouteEntry for non-keyed items and records them for
+    //       transitionDelegate.
+    //     * Sync keyed items with the source if it exists.
+    // 5. Walk the the narrowed part of the old list again to records the
```
remove one the.

goderbauer commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -1991,8 +2653,276 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
+    //       transitionDelegate.
+    //     * Sync keyed items with the source if it exists.
+    // 5. Walk the the narrowed part of the old list again to records the
+    //    _RouteEntry[s] ,as well as pageless routes, needed to be removed for
```
nit: move whitespace after the comma

goderbauer commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -1991,8 +2653,276 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
+    //    transitionDelegate.
+    // 5. Walk the bottom of the list again, syncing the nodes and recording
+    //    pageless routes.
+    // 6. Use transitionDelegate to for explicit decisions on how _RouteEntry[s]
```
remove the to.

goderbauer commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -505,6 +620,271 @@ class NavigatorObserver {
+    return results;
+  }
+
+  /// A method that will be called by the [Navigator] for deciding how route
```
.. to decide how routes transition in or out of the screen when [Navigator.pages] is updated.

goderbauer commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -505,6 +620,271 @@ class NavigatorObserver {
+  /// A method that will be called by the [Navigator] for deciding how route
+  /// transitions in or out of the screen.
+  ///
+  /// The [newPageRouteHistory] represents the new page-based route after the
```
newPageRouteHistory (and all other method parameters mentioned in this paragraph should be in ``.

goderbauer replied:
The `newPageRouteHistory` list contains all page-based routes in order that will be on the [Navigator]'s history stack after this update completes.

goderbauer commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -505,6 +620,271 @@ class NavigatorObserver {
+  /// The [newPageRouteHistory] represents the new page-based route after the
+  /// pages update. If a route in [newPageRouteHistory] has its
+  /// [StageableRoute.isEntering] set to true, this route requires explicit
+  /// decision on how it transition in the screen. To make a decision, use
```
how it should transition onto the Navigator.

goderbauer replied:
use -> call?

goderbauer commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -505,6 +620,271 @@ class NavigatorObserver {
+  /// history before the update. The key location is the page-based route
+  /// directly below the removed route. The key location is null if the route to
+  /// be removed is the bottom most route. To make a decision for a removed
+  /// route, use [StageableRoute.markForPop], [StageableRoute.markForComplete]
```
use -> call?

goderbauer commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -505,6 +620,271 @@ class NavigatorObserver {
+  ///
+  /// Once all the decisions have been made, this method must merge the removed
+  /// routes and the [newPageRouteHistory] and return the merged result. The
+  /// order in the result will be the order the [Navigator] used for updating
```
used -> uses

chunhtai commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -1952,6 +2423,10 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
   @override
   void didUpdateWidget(Navigator oldWidget) {
     super.didUpdateWidget(oldWidget);
+    assert(
```
I can't avoid the assert here, please see https://github.com/flutter/flutter/pull/50362/files#r389104877

chunhtai committed:
addressing pr comment

chunhtai commented:
@goderbauer I addressed all the comments. I also rename `PageBuilder` to `CustomBuilderPage` because `PageBuilder` implies a builder that returns a page which is a bit weird.

chunhtai commented:
bump for attention

chunhtai committed:
fix some sentences

chunhtai committed:
fix typo

Hixie commented on dev/integration_tests/android_views/lib/main.dart:
```diff
@@ -7,7 +7,7 @@ import 'package:flutter_driver/driver_extension.dart';
 import 'motion_events_page.dart';
 import 'page.dart';
 
-final List<Page> _allPages = <Page>[
```
can we do this change in a separate PR to simplify this one?

Hixie commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -50,6 +55,15 @@ typedef RoutePredicate = bool Function(Route<dynamic> route);
+/// is passed into [Navigator.pages] so that it no longer includes the
+/// corresponding [Page]. (Otherwise, the page will be interpreted as a new page
+/// to show when the [Navigator.pages] list is next updated.)
+typedef OnPopPageCallback = bool Function(Route<dynamic> route, dynamic result);
```
please use the naming scheme described in our style guide

Hixie commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -247,7 +287,6 @@ abstract class Route<T> {
   ///
   /// See [popped], [didComplete], and [currentResult] for a discussion of the
   /// `result` argument.
-  @protected
```
how come this is no longer protected but didComplete is?

Hixie commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -461,6 +500,82 @@ class RouteSettings {
+  /// [key].
+  bool canUpdate(Page<dynamic> other) {
+    return other.runtimeType == runtimeType &&
+      other.key == key;
```
trivial nit: indenting here is off

Hixie commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -505,6 +620,273 @@ class NavigatorObserver {
 
+/// A [Route] wrapper interface that can be staged for [TransitionDelegate] to
+/// decide how its underlying [Route] should transition on or off screen.
+abstract class StageableRoute{
```
trivial nit: space missing before {

Hixie commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -505,6 +620,273 @@ class NavigatorObserver {
+    //   popped, completed or removed.
+    // - All the entering routes in the result must preserve the same order as
+    //   the entering routes in newPageRouteHistory, and the result must contain
+    //   all exiting routes.
```
are these rules documented in the API docs somewhere?

Hixie commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -505,6 +620,273 @@ class NavigatorObserver {
+
+      assert(
+       indexOfNextRouteInNewHistory == newPageRouteHistory.length &&
+       exitingPageRoutes.isEmpty
```
indentation is off

Hixie commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -505,6 +620,273 @@ class NavigatorObserver {
+  /// order in the result will be the order the [Navigator] uses for updating
+  /// the route history. The return list must preserve the same order of routes
+  /// in `newPageRouteHistory`. The removed routes, however, can be inserted
+  /// into the return list freely as long as all of them are included.
```
these docs are great

Hixie commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -505,6 +620,273 @@ class NavigatorObserver {
+  ///    it exit the screen without an animated transition.
+  ///  * [DefaultTransitionDelegate.resolve], which implements the default way
+  ///    to decide how routes transition in or out of the screen.
+  Iterable<StageableRoute> resolve({
```
to validate this API we should include some examples of different transition delegates that do very different things from each other. maybe enumerate some use cases for this and implement two that are on the opposite end of the spectrum.

Hixie commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -701,12 +1099,12 @@ class NavigatorObserver {
-/// because it doesn't depend on its animation parameters (elided with `_`
-/// and `__` in this example). The transition is built on every frame
-/// for its duration.
+/// The route is built in two parts, the "page widget" and the "transitions". The
```
Page isn't a widget, is it?

Hixie replied:
we should rewrite this without using the word "page". Maybe "body".

Hixie replied:
i guess since we call it a pageBuilder that would be hard. hrm.

Maybe we can deprecated pageBuilder and rename it bodyBuilder and have pageBuilder just forward to bodyBuilder (assert only one is non-null)

goderbauer replied:
We have lots of things that use the word Page, but they are not related to a Page in the Navigator 2.0 sense, e.g. PageRouteBuilder, MaterialPageRoute, PageRoute, etc. All of this is as confusing as the pageBuilder property here. Re-naming all of these seems like a pretty big API churn. Maybe we should instead find a different word for page in the new API and leave the existing things as-is?

Hixie replied:
What term can we use instead?

goderbauer replied:
I don't have a great idea. Maybe we just leave things how they are for now.

chunhtai replied:
I don't have anything concrete neither, It seems like the right thing to do is to migrate all other classes that has name 'Page' in it because they are named after page for no reason and now we have a thing that can only be named 'Page' and nothing else. I agree we should leave it as is and migrate them if developers start to complain about the name is confusing with other classes.

chunhtai committed:
addressed some coments

chunhtai committed:
add an example to transition delegate doc string

chunhtai commented:
I addressed all comments except for this one https://github.com/flutter/flutter/pull/50362#discussion_r398223925. @goderbauer @Hixie

goderbauer commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -491,6 +606,330 @@ class NavigatorObserver {
 
+/// A [Route] wrapper interface that can be staged for [TransitionDelegate] to
+/// decide how its underlying [Route] should transition on or off screen.
+abstract class RouteTransitionRecord {
```
I like this new name :)

goderbauer commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -859,17 +1314,75 @@ class NavigatorObserver {
   ///
-  /// The [onGenerateRoute] argument must not be null.
+  /// The [onGenerateRoute], [pages], [onGenerateInitialRoutes],
+  /// [transitionDelegate], [observers]  arguments must not be null.
```
document under what conditions you have to provide the onPopPage callback?

goderbauer commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -1895,17 +2469,54 @@ class _RouteEntry {
+
+  @override
+  void markForPush() {
+    assert(isEntering && !_debugWaitingForExitDecision);
```
The asserts in the mark* methods should have an error text telling people what's wrong.

goderbauer reviewed (approved):
LGTM

chunhtai commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -50,6 +55,15 @@ typedef RoutePredicate = bool Function(Route<dynamic> route);
 
+/// Signature for the [Navigator.onPopPage] callback.
+///
+/// This callback must call [Route.didPop] or [Route.didComplete] on the
```
oops, nice catch !

chunhtai committed:
addressing coments

chunhtai committed:
fix doc
